src/tools/imgtool/charconv.cpp

```diff
@@ -2,161 +2,125 @@
 // copyright-holders:Nathan Woods
 /***************************************************************************
 
-    charconv.c
+    charconv.cpp
 
     Imgtool character set conversion routines.
 
 ***************************************************************************/
 
 #include "corestr.h"
 #include "charconv.h"
+#include "unicode.h"
+#include "coretmpl.h"
 
+imgtool::simple_charconverter imgtool::charconverter_iso_8859_1(nullptr, nullptr);
 
-/*-------------------------------------------------
-    utf8_from_latin1 - convert an ISO-8859-1
-    character sequence to an UTF-8 string
--------------------------------------------------*/
 
-static char *utf8_from_latin1(const char *src)
-{
-	char *buffer, *bufptr;
+//-------------------------------------------------
+//  from_utf8
+//-------------------------------------------------
 
-	/* validate input */
-	if (!src)
-	{
-		return nullptr;
-	}
-
-	/* allocate space for result, twice the source len to be safe */
-	buffer = (char *) malloc(strlen(src) * 2 + 1);
+void imgtool::charconverter::from_utf8(std::ostream &dest, const std::string &src) const
+{
+	from_utf8(dest, src.c_str(), src.size());
+}
 
-	/* point to the start */
-	bufptr = buffer;
 
-	do
-	{
-		unsigned char c = *src;
+//-------------------------------------------------
+//  to_utf8
+//-------------------------------------------------
 
-		if (c < 0x80)
-		{
-			*bufptr++ = c;
-		}
-		else if (c < 0xc0)
-		{
-			*bufptr++ = '\xc2';
-			*bufptr++ = c;
-		}
-		else
-		{
-			*bufptr++ = '\xc3';
-			*bufptr++ = c - 0x40;
-		}
-	} while (*src++);
-
-	return buffer;
+void imgtool::charconverter::to_utf8(std::ostream &dest, const std::string &src) const
+{
+	to_utf8(dest, src.c_str(), src.size());
 }
 
 
-/*-------------------------------------------------
-    latin1_from_utf8 - convert an UTF-8
-    character sequence to an ISO-8859-1 string
--------------------------------------------------*/
+//-------------------------------------------------
+//  simple_charconverter::simple_charconverter
+//-------------------------------------------------
 
-static char *latin1_from_utf8(const char *src)
+imgtool::simple_charconverter::simple_charconverter(const char32_t lowpage[0x80], const char32_t highpage[0x80], unicode_normalization_form norm)
+	: m_norm(norm), m_lowpage(lowpage), m_highpage(highpage)
 {
-	char *buffer, *bufptr;
-
-	/* validate input */
-	if (!src)
+	// build the reverse lookup table
+	for (int i = 0; i < 256; i++)
 	{
-		return nullptr;
+		const char32_t *page = i >= 128 ? m_highpage : m_lowpage;
+		char32_t unicode_char = page ? page[i % 128] : i;
+		m_reverse_lookup.emplace_back(unicode_char, (char)i);
 	}
 
-	/* allocate space for result */
-	buffer = (char *) malloc(strlen(src) + 1);
-
-	/* point to the start */
-	bufptr = buffer;
-
-	do
+	// and sort it
+	std::sort(m_reverse_lookup.begin(), m_reverse_lookup.end(), [](const std::pair<char32_t, char> &a, const std::pair<char32_t, char> &b)
```

Comment:
I don't believe that I'm doing this twice.  I do have another comparison roughly similar, but that scenario the second parameter is 'const char32_t &', not 'const std::pair<char32_t, char>'